pkg/kgo/compression.go

```diff
@@ -266,15 +268,25 @@ type zstdDecoder struct {
 	inner *zstd.Decoder
 }
 
-func (d *decompressor) decompress(src []byte, codec byte) ([]byte, error) {
+func (d *decompressor) decompress(src []byte, codec byte, pool *pool.BucketedPool[byte]) ([]byte, error) {
 	// Early return in case there is no compression
 	compCodec := codecType(codec)
 	if compCodec == codecNone {
 		return src, nil
 	}
-	out := byteBuffers.Get().(*bytes.Buffer)
-	out.Reset()
-	defer byteBuffers.Put(out)
+	var out *bytes.Buffer
+
+	if pool != nil {
+		outBuf := pool.Get(pool.MaxSize())[:0]
```

Comment:
we can figure out the length of the message for most compression algos (if not all). should we try that instead of allocating massive buffers just in case?

Reply:
my understanding was that it wasn't something so trivial. but if you can point me to a reference on how to do it, I would definitely prefer that option. 👍

Reply:
here's a proposal that came to mind. let me know what you think.

```go
	// Make a guess at the output size.
	decodedBufSize := len(src) * 2
	if compCodec == codecSnappy {
		var err error
		decodedBufSize, err = s2.DecodedLen(src)
		if err != nil {
			return nil, err
		}
	}

	outBuf := pool.Get(decodedBufSize)[:0]
	defer func() {
		pool.Put(outBuf)
	}()
	out = bytes.NewBuffer(outBuf)
```

Reply:
_changed @ 2df6656ebf28adc3477b5d75529371e37a7a8a1f_

Reply:
yeah i had something similar in mind. Most of the decoders we use expose methods for the length of the decoded blob. Some don't, like gzip i think, but then we can just fall back to the `len(src)*2` (which is still better imo than `pool.MaxSize()`)